### ISSUE TITLE
shared variable and threads

### ISSUE THREAD

**Issue by deepj** [oracle/truffleruby#1548]
Honestly, I'm not sure how to describe this error by words (too late at night here)

I noticed this in `connection_pool` tests => https://github.com/mperham/connection_pool/blob/master/test/test_connection_pool.rb#L144

To reproduce:

```ruby
require 'timeout'
require 'connection_pool'

pool = ConnectionPool.new(timeout: 0, size: 1) { Object.new }
def pool.checkout(options)
  sleep 0.015
  super
end

did_something = false
begin
  Timeout.timeout(0.01) do
    pool.with do |obj|
      did_something = true
      # Timeout::Error will be triggered by any non-trivial Ruby code
      # executed here since it couldn't be raised during checkout.
      # It looks like setting the local variable above does not trigger
      # the Timeout check in MRI 2.2.1.
      obj.tap { obj.hash }
    end
  end
rescue Timeout::Error
  # Nothing
end

puts did_something
```

Error is `did_something` is `false`, but would be `true` such under MRI

**Comment by chrisseaton** [comment]
This seems to be down to some kind of difference about when interrupts can be handled. Notice that it's commented that JRuby also fails this test (but actually it does work with your code above). Rubinius 3.107 also fails.

I doubt it's anything to do with the scope of variables, as suggested in your title.

`handle_interrupt` is very subtle and complicated and I'm sure the error is in there.

Here's a version that doesn't need `connection_pool` installed.

```ruby
require 'timeout'

def checkout
  sleep 2
end

def checkin
end

def with
  Thread.handle_interrupt(Exception => :never) do
    checkout
    begin
      Thread.handle_interrupt(Exception => :immediate) do
        yield
      end
    ensure
      checkin
    end
  end
end

reached_here = false

begin
  Timeout.timeout(1) do
    with do
      reached_here = true
    end
  end
rescue
  # nothing
end

p reached_here
```

Works on MRI, JRuby. Fails on Rubinius, TruffleRuby.

`Timeout` uses `Thread#raise`, which uses the `thread_raise` primitive, which uses a guest-language safe point to run an action in the target thread. That action is run in the target thread whenever it calls `poll`. It calls `poll`. We can look at which call to `poll` is seeing the exception.

It's being interrupted at `Thread.handle_interrupt(Exception => :immediate) do`.

That seems to make sense to me - it's already timed out, so as soon as we re-enable interrupts it is interrupted. We aren't even explicitly testing for an interrupt after re-enabling, which would make sense to me as well, it's just naturally happening because we've re-enabled and are running again.

Maybe @mperham would have any time to comment on what he thinks about that? Why does he expect the program to keep running all the way until the point of being half-way through the yielded block before it realises it's been interrupted?

I think the spec is really about the object being returned to the pool despite it timing out while being checked out, and it is - that bit works.

**Comment by eregon** [comment]
Right, that makes sense to me. In TruffleRuby a block/Proc checks safepoints before executing the block/Proc and so here it just triggers when entering the block given to `Thread.handle_interrupt`, before the `yield`:
```
$ bin/ruby s.rb                                                                  
s.rb:14:in `handle_interrupt': execution expired (Timeout::Error)
	from s.rb:14:in `handle_interrupt'
	from s.rb:14:in `block in with'
	from s.rb:11:in `handle_interrupt'
	from s.rb:11:in `handle_interrupt'
	from s.rb:11:in `with'
	from s.rb:27:in `block in <main>'
	from truffleruby/lib/truffle/timeout.rb:164:in `timeout'
	from s.rb:26:in `<main>'
```

Safepoints / interrupt points polling places are implementation-specific.
IMHO in this case it is a test problem. It would work fine if the test checks that the connection is checked back in. It would also be useful to ensure a connection is actually checked out, which could be done by e.g., mocking `checkout` (and `checkin`).

**Comment by mperham** [comment]
I would be happy to see a better version of the test.  I got something working in MRI 2.2 and haven't touched it since.  As noted, it's been flaky across Ruby impls.

**Comment by deepj** [comment]
@chrisseaton what a title of this issue would be more accurate here. I have no idea.

**Comment by chrisseaton** [comment]
We've documented that TruffleRuby detects thread interrupts in different places than MRI. This is the same as Rubinius and JRuby, and TruffleRuby actually detects them sooner, rather than later, so I don't see it as a problem.

https://github.com/oracle/truffleruby/blob/56e619506e7265f34f4e981e9072ed902f2b1e53/doc/user/compatibility.md#threads-detect-interrupts-at-different-points

So I'm going to close this as won't-fix, but I'll send `connection_pool` a PR for their spec first.

**Comment by chrisseaton** [comment]
https://github.com/mperham/connection_pool/pull/112
